src/hera/workflows/resource_base.py

```diff
@@ -0,0 +1,197 @@
+from hera.shared._pydantic import PydanticBaseModel, get_fields
+
+
+class ModelMapper:
+    """Maps a model attribute to the attribute path of the produced resource.
+
+    For example `foo: MopdelMapper("metadata.name")` would map the annotated object's
+    `foo` attribute to the "metadata" -> "name" field on the final k8s resource.
+    """
+
+    def __init__(self, model_path: str, hera_builder: Optional[Callable] = None):
+        """Accept the `model_path` to map to."""
+        self.model_path = None
+        self.builder = hera_builder
+
+        if not model_path:
+            # Allows overriding parent attribute annotations to remove the mapping
+            return
+
+        self.model_path = model_path.split(".")
+
+    @classmethod
+    def build_model(
+        cls,
+        hera_class: Type[ResourceBase],
+        hera_obj: ResourceBase,
+        model: PydanticBaseModel,
+        *,
+        traverse_mro: bool = True,
+    ):
+        """Builds an instance of a `ResourceBase` into the destination model type."""
+        assert isinstance(hera_obj, ResourceBase), type(hera_obj)
+
+        for attr, mapper, _ in hera_class._iter_model_mappers(traverse_mro=traverse_mro):
+            # Value comes from builder function if it exists on hera_obj, otherwise directly from the attr
+            value = (
+                getattr(hera_obj, mapper.builder.__name__)() if mapper.builder is not None else getattr(hera_obj, attr)
+            )
+            if value is not None:
+                _set_model_attr(model, mapper.model_path, value)
+
+        return model
+
+
+class ResourceBase(BaseMixin):
+    """Base class for top-level Hera CRDs.
+
+    Subclasses should supply a `mapped_model` attribute, which is the Pydantic model
+    which represents the final k8s resource.
+
+    Note, this class will validate that the annotated attributes map to real fields on the `mapped_model`.
+    """
+
+    mapped_model: ClassVar[Type[PydanticBaseModel]]
+
+    def __init_subclass__(cls, *, traverse_mro=True, **kwargs):
+        """Validate that the annotated `ModelMapper` paths map to real destination fields."""
+        for _, model_mapper, _ in cls._iter_model_mappers(traverse_mro=traverse_mro):
+            curr_cls = cls.mapped_model
+
+            for key in model_mapper.model_path:
+                fields = get_fields(curr_cls)
+                if key not in fields:
+                    raise ValueError(f"Model key '{key}' does not exist in class {curr_cls.__qualname__}")
+                curr_cls = fields[key].outer_type_
+
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def _iter_model_mappers(cls, traverse_mro: bool = True):
+        mro = cls.__mro__
+        if not traverse_mro:
+            mro = (cls,)
+
+        # Due to child subclasses being able to override parent fields, traversing the mro might
+        # yield the same attribute more than once, which wont make sense. Track seen fields, and only
```

Comment:
I think this problem was solved using a `ChainMap` but I don't remember exactly

Reply:
jk, done now. it **does** simplify the impl back down. thanks for pointing that out!